chef/cookbooks/ceph/providers/client.rb

```diff
@@ -64,23 +64,20 @@ def auth_set_key(ceph_conf, admin_keyring, keyname, caps)
   cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
   get_or_create = Mixlib::ShellOut.new(cmd)
   get_or_create.run_command
-  unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/).empty?
-    Chef::Log.info("Deleting old key with incorrect caps")
-    # don't remove custom existing caps for osd pools
-    cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
-    if cur_caps.empty?
-      caps_osd = caps["osd"]
-    else
-      caps_osd = cur_caps["osd"] + "," + caps["osd"]
-    end
-    caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(",")
-    caps_str = caps.map { |k, v| "#{k} '#{v}'" }.join(" ")
-    cmd = "ceph -k #{admin_keyring} -c #{ceph_conf} auth get-or-create #{keyname} #{caps_str}"
-    # delete an old key if it exists and is wrong
-    Mixlib::ShellOut.new("ceph -k #{admin_keyring} -c #{ceph_conf} auth del #{keyname}").run_command
-    # try to create again
-    get_or_create = Mixlib::ShellOut.new(cmd)
-    get_or_create.run_command
+  return get_or_create.error! unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/)
+  Chef::Log.info("Updating incorrect caps for #{keyname}")
+  # don't remove custom existing caps for osd pools
+  cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
+  if cur_caps.empty?
+    caps_osd = caps["osd"]
+  else
+    caps_osd = cur_caps["osd"] + "," + caps["osd"]
   end
-  get_or_create.error!
+  caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(", ")
```

Comment:
shouldn't that unify based only on the first word rather than the complete substring?

if you have something like



images ro,images rw

this wouldn't unify. I thought what the logic should do is overwrite the permission values with what has been provided by the function parameter. here you're only preserving attributes for completely unrelated pools.